### DISCUSSION TITLE
Pack a box up to a certain volume?

### DISCUSSION THREAD

**Discussion by eastpiada** [dvdoug/BoxPacker#216] [Q&A] [answered]
Was wondering if this project has a way to pack a box up to a certain volume or volume percentage given the inner dimensions of a box? For example, I have a box with inner dimensions (inches) of 12x10x8 and I would like to fill it up to 80% of it's volume capacity to leave room for filler.

Looking through the code it doesn't seem explicitly available, but there are methods available (usedVolume, volumeUtilization) where maybe this could be done in a round about way?

**Comment by dvdoug** [comment, marked as answer]
Hello

There's no built-in mechanism for doing something like that, but depending on how precise you want to be, either of the following should be feasible.

1) Reduce the "inner" dimensions of each box by a couple of centimetres to ensure space for the additional protective packaging
2) You could also do something with the custom constraint system (https://www.boxpacker.io/en/stable/advanced-usage.html#custom-constraints) - that doesn't give you the `PackedBox` and its helper methods (I'll look into changing that at the next major version bump), but you should be able to work out the current packed volume anyway to make a decision with

**Reply by eastpiada** [reply]
Thanks Doug for a quick response and a well maintained and amazing package!

Thinking through option 1, let's say I have 3 available box sizes, small, medium and large. I need to pack item A which has its longest dimension that won't fit the small box, but will fit the medium box exactly. What I don't want to do is reduce the inner dimensions of the medium box such that the longest dimension of item A now won't fit the medium box just to accommodate for a volume limitation. Am I thinking about this correctly? I don't want to push item A to a large box just because I reduced the inner dims of the medium box as that seems inefficient. If I could fit up to 5 of item A into a medium box based on the original inner dims, lets say by volume I would only want to allow 4 of item A due to my volume limitation.

I'll have to think a little more about option 2, but I think I understand where you're going with that...

**Reply by dvdoug** [reply]
> I need to pack item A which has its longest dimension that won't fit the small box, but will fit the medium box exactly.

A thought experiment here - if that scenario is OK where *some* edges/sides of your items are not protected by additional filler but only by the shipping box, then why not is a scenario where none of them are🤔 ? There's no right or wrong answer there of course, it depends on what you're shipping...

> I'll have to think a little more about option 2, but I think I understand where you're going with that...

I can't change that interface without a major version bump, but I can add a `getVolume()` helper to `PackedItemList`. So something like this should work.

```php

class YourItem implements ConstrainedPlacementItem {

public function canBePacked(
            Box $box,
            PackedItemList $alreadyPackedItems,
            int $proposedX,
            int $proposedY,
            int $proposedZ,
            int $width,
            int $length,
            int $depth
        ) {
            $boxVolume = $box->getInnerWidth() * $box->getInnerLength() * $box->getInnerDepth();
            return ($alreadyPackedItems->getVolume() + $this->getVolume()) / $boxVolume <= 0.8;
        }
}

```

**Reply by eastpiada** [reply]
> A thought experiment here - if that scenario is OK where some edges/sides of your items are not protected by additional filler but only by the shipping box, then why not is a scenario where none of them are🤔 ? There's no right or wrong answer there of course, it depends on what you're shipping...

In my world it is okay for one or more item sides to be protected by nothing but the shipping box itself. For larger items where placement in the box is physically easy to do, a dimensional fit would be fine. However, we have use cases where the item being picked might be a small 1x1x1 (inch) box and the picker is just dumping a large qty of them into the small/medium/large shipping box and not neatly stacking them (because they know the full inventory box on the shelf for the 1x1x1 item has 20 in it so they just dump them all in). In this scenario I'd need to check how much remaining volume I have available because of the less than efficient packing job. We have for years now used pretty much a volume only algorithm and so our people have gotten used to a "fluff factor". I would definitely like to get to a point where I can reduce this fluff factor, but we'll still need a bit of it. Over committing too many items to a chosen shipping box definitely doesn't make our operational managers happy 😁

> I can't change that interface without a major version bump, but I can add a getVolume() helper to PackedItemList. So something like this should work.

I think this is a great idea. Seems I'll need to make all my items implement the ConstrainedPlacementItem interface, but I don't guess that'll cause too much overhead.

**Reply by dvdoug** [reply]
https://github.com/dvdoug/BoxPacker/releases/tag/3.7.0
